src/main/scala/io/iohk/ethereum/blockchain/sync/RegularSyncController.scala

```diff
@@ -0,0 +1,193 @@
+    val parentByNumber = blockchain.getBlockHeaderByNumber(headers.head.number - 1)
+
+    parentByNumber match {
+      case Some(parent) if checkHeaders(headers) =>
+        //we have same chain
+        if (parent.hash == headers.head.parentHash) {
+          sendWithTimeout(peer, SendMessage(GetBlockBodies(headers.take(blockBodiesPerRequest).map(_.hash))))
+        } else {
+          self ! ResolveBranch(peer)
+        }
+      case _ =>
+        log.warning("got header that does not have parent")
+        resumeWithDifferentPeer()
+    }
+  }
+
+  private def handleBlockBodies(m: BlockBodies) = {
+    if (m.bodies.nonEmpty) {
+      val result = headersQueue.zip(m.bodies).map { case (h, b) => blockValidator(h, b) }
+
+      if (!result.exists(_.isLeft)) {
+        val blocks = result.collect { case Right(b) => b }
+
+        blockchain.getBlockHeaderByHash(blocks.head.header.parentHash)
+          .flatMap(b => blockchain.getTotalDifficultyByHash(b.hash)) match {
+          case Some(td) =>
+            var currentTd = td
+            val newBlocks = blocks.map { b =>
+              val blockHashToDelete = blockchain.getBlockHeaderByNumber(b.header.number).map(_.hash).filter(_ != b.header.hash)
+              blockchain.save(b)
+              appStateStorage.putBestBlockNumber(b.header.number)
+              currentTd += b.header.difficulty
+              blockchain.save(b.header.hash, currentTd)
+              blockHashToDelete match {
+                case Some(hash) => blockchain.removeBlock(hash)
+                case _ =>
+              }
+
+              NewBlock(b, currentTd)
+            }
+
+            self ! BroadcastBlocks(newBlocks)
+            log.info(s"got new blocks up till block: ${newBlocks.last.block.header.number} " +
+              s"with hash ${Hex.toHexString(newBlocks.last.block.header.hash.toArray[Byte])}")
+          case None =>
+            log.error("no total difficulty for latest block")
+        }
+
+        headersQueue = headersQueue.drop(result.length)
+        if (headersQueue.nonEmpty) {
+          sender() ! SendMessage(GetBlockBodies(headersQueue.take(blockBodiesPerRequest).map(_.hash)))
+        } else {
+          resolvingBranch = false
+          self ! ResumeRegularSync
+        }
+      } else {
+        //blacklist for errors in blocks
+        resumeWithDifferentPeer()
+      }
+    } else {
+      //we got empty response for bodies from peer but we got block headers earlier
+      resumeWithDifferentPeer()
+    }
+  }
+
+  private def sendWithTimeout(peer: ActorRef, m: SendMessage[_]) = {
+    peer ! m
+    timeout = Some(scheduler.scheduleOnce(peerResponseTimeout, self, PeerTimeOut(peer)))
+  }
+
+  private def cancelTimeout() = {
+    timeout.foreach(_.cancel())
+    timeout = None
+  }
+
+  private def scheduleResume() = {
+    headersQueue = Seq.empty
+    scheduler.scheduleOnce(checkForNewBlockInterval, self, ResumeRegularSync)
+  }
+
+  private def resumeWithDifferentPeer() = {
```

Comment:
Maybe we need set `resolvingBranch` to false here as well. In the scenario that we are resolving a branch, we get an unexpected response and we have to call `resumeWithDifferentPeer` in `handleBlockBranchResolution`, shouldn't `resolvingBranch` be set to false so that the next `BlockHeaders` are handled properly?